library/core/src/mem/drop_guard.rs

```diff
@@ -0,0 +1,166 @@
+    /// ```
+    #[unstable(feature = "drop_guard", issue = "none")]
+    pub fn new(inner: T, f: F) -> Self {
+        Self { inner: ManuallyDrop::new(inner), f: ManuallyDrop::new(f) }
+    }
+
+    /// Consumes the `Guard`, returning the wrapped value.
+    ///
+    /// This will not execute the closure. This is implemented as a static
+    /// method to prevent any potential conflicts with any other methods called
+    /// `into_inner` from the `Deref` and `DerefMut` impls.
+    ///
+    /// It is typically preferred to call this function instead of `mem::forget`
+    /// because it will appropriately de-initialize the closure stored in the
+    /// `DropGuard` before returning the stored value.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # #![allow(unused)]
+    /// use std::mem::DropGuard;
+    ///
+    /// let value = String::from("Nori likes chicken");
+    /// let guard = DropGuard::new(value, |s| println!("{s}"));
+    /// assert_eq!(Guard::into_inner(guard), "Nori likes chicken");
+    /// ```
+    #[unstable(feature = "drop_guard", issue = "none")]
+    #[inline]
+    pub fn into_inner(guard: Self) -> T {
+        // Implementation note: this is the same approach that scopeguard takes.
+
+        // First we ensure that dropping the guard will not trigger
+        // its destructor
+        let mut guard = ManuallyDrop::new(guard);
+
+        // Next we manually read the stored value from the guard.
+        //
+        // SAFETY: this is safe because we've taken ownership of the guard.
+        let value = unsafe { ptr::read(&*guard.inner) };
+
+        // Finally we drop the stored closure. We do this *after* having read
+        // the value, so that even if the closure's `drop` function panics,
+        // unwinding still tries to drop the value.
+        //
+        // SAFETY: this is safe because we've taken ownership of the guard.
+        unsafe { ManuallyDrop::drop(&mut guard.f) };
+        value
+    }
+}
+
+#[unstable(feature = "drop_guard", issue = "none")]
+impl<T, F> Deref for DropGuard<T, F>
+where
+    F: FnOnce(T),
+{
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &*self.inner
+    }
+}
+
+#[unstable(feature = "drop_guard", issue = "none")]
+impl<T, F> DerefMut for DropGuard<T, F>
+where
+    F: FnOnce(T),
+{
+    fn deref_mut(&mut self) -> &mut T {
+        &mut *self.inner
+    }
+}
+
+#[unstable(feature = "drop_guard", issue = "none")]
+impl<T, F> Drop for DropGuard<T, F>
+where
+    F: FnOnce(T),
+{
+    fn drop(&mut self) {
+        // SAFETY: we're taking the values out of the `ManuallyDrop` with
+        // the express intent of dropping them.
+        let inner = unsafe { ManuallyDrop::take(&mut self.inner) };
+        let f = unsafe { ManuallyDrop::take(&mut self.f) };
+        f(inner);
+    }
+}
+
+// tests copied from https://docs.rs/scopeguard/latest/src/scopeguard/lib.rs.html#1-595
+#[cfg(test)]
+mod tests {
```

Comment:
Maybe add a test for panics occurring within the closure since `into_inner` takes extra precaution to get it right?

Reply:
I'm not sure what that test would look like, since I'm not sure what the current panic-while-unwinding behavior is supposed to be. How would you write it?

Reply:
`#[test]` inside `library/core` is ignored. Tests for libcore should be in `library/coretests` instead, which can use libstd just fine.